### PR TITLE
[CFG] text and/or audio classifier-free-guidance training and inference recipe

### PR DESCRIPTION
**Pull request by XuesongYang**
1. [Text-CFG] text classifier-free-guidance training and inference recipe, and option of applying cfg-filter further to address fast speech rates.
    - [inference] glue conditioned and unconditioned batches, and interpolate their logits. Next token was sampled from the interpolated distribution.
    - [inference with cfg-filter] obtained a top-k mask given interpolated logits and apply the mask to its conditioned logits counterpart. It also provide an option of interpolating the masked conditioned logits with original unconditioned logits furthermore.
2. [Text/Audio-CFG] text cfg and/or audio cfg training and inference recipe.
    - [training] added support of randomly dropping either transcripts or audio context or both in a batch. 
    - [inference] glue conditioned and unconditioned batches and interpolate their logits. Next token was sampled from the interpolated distribution. Unconditioned batches could be relevant to either transcript only, or audio context only, or both. 
    - [inference with cfg-filter] obtained a top-k mask given interpolated logits and apply the mask to its conditioned logits counterpart. It also provide an option of interpolating the masked conditioned logits with original unconditioned logits furthermore.
3. added comments to some variables.
4. refined some debugging info.

# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Jenkins CI
To run Jenkins, a NeMo User with write access must comment `jenkins` on the PR.

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py [hunk 1]
```diff
@@ -725,6 +725,7 @@ def __init__(self, *args, **kwargs):
         self.speech_head_type = "token_level"
         self.attn_prior_scaledown_start_step = 10000
         self.attn_prior_end_step = 11000
+        self.use_alignment_loss = False
```
Comment (blisc) [inline nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py:729]: Why does this default to False?

File: nemo/collections/tts/models/speechllm/megatron_t5_speechllm_model.py [hunk 2]
```diff
@@ -801,6 +833,40 @@ def on_validation_epoch_start(self) -> None:
     def training_step(self, dataloader_iter, batch_idx):
         self._optimizer.zero_grad()
         batch = next(dataloader_iter)
+
+        # apply classifier-free guidance by replacing input question tokens with [UNK].
+        if self._rng.random() < self.question_free_guidance_prob:
```
Comment (blisc) [inline nemo/collections/tts/models/speechllm/megatron_t5_speechllm_model.py]: Can we skip `self._rng.random()` if tranining with cfg is disabled?

File: nemo/collections/tts/models/speechllm/megatron_t5_speechllm_model.py [hunk 3]
```diff
@@ -1432,25 +1498,55 @@ def predict_step(self, batch: Any, batch_idx: int, dataloader_idx: int = 0, log_
...
+            # TODO @xueyang: potential bug when max_inference_timesteps < dec_input.shape[2], then dec_input is clipped.
             dec_input = torch.nn.functional.pad(dec_input, (0, max_inference_timesteps - dec_input.shape[2]), value=0)
             dec_input[:, :, self.decoder_context_len + 1:].zero_()
+            # TODO @xueyang: why not just declare torch.ones(dec_input_raw.size(0), max_inference_timesteps)?
             dec_input_mask = torch.nn.functional.pad(
                 dec_input_mask, (0, max_inference_timesteps - dec_input_mask.shape[1]), value=1
             )
 
+            # question unconditioned case
+            if self.question_guidance_scale != 1.0:
```
Comment (blisc) [inline nemo/collections/tts/models/speechllm/megatron_t5_speechllm_model.py]: Can you add an inference cfg parameter or option to the predict step?

**Review by blisc — CHANGES_REQUESTED** [review]

File: nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py (hunk 1 above)
Comment (XuesongYang) [reply nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py:729]: i tried to keep consistent defaults by following the same default setup at `Line 201:nemo/collections/tts/models/speechllm/megatron_t5_speechllm_model.py`. 

```python
if cfg.get('use_alignment_loss', False):
```

File: nemo/collections/tts/models/speechllm/megatron_t5_speechllm_model.py (hunk 2 above)
Comment (XuesongYang) [reply nemo/collections/tts/models/speechllm/megatron_t5_speechllm_model.py]: yeah, better to skip `self._rng.random()` at each training step when cfg is disabled to save some computations. Let me add something like below,
```suggestion
        if self.question_free_guidance_prob > 0.0:
            if self._rng.random() < self.question_free_guidance_prob:
                ...
            else:
                # recover to original alignment loss config.
                self.frozen_model.enc_dec_model.use_alignment_loss = self.use_alignment_loss
```

File: nemo/collections/tts/models/speechllm/megatron_t5_speechllm_model.py (hunk 3 above)
Comment (XuesongYang) [reply nemo/collections/tts/models/speechllm/megatron_t5_speechllm_model.py]: sure thing. Actually, I've already introduced `apply_text_cfg` and `apply_audio_cfg` params in my text and/or audio cfg updates. Something like below. Will refactor this PR a little bit to combine both text and audio cfg together since minor changes added for audio cfg.

```suggestion
            # text unconditioned case
            if self.apply_text_cfg:
                ....
            
            # audio unconditioned case
            if self.apply_audio_cfg:
                ...
```

Comment (XuesongYang) [reply nemo/collections/tts/models/speechllm/megatron_t5_speechllm_model.py:1524]: code snippet is here: https://gitlab-master.nvidia.com/xueyang/nemo/-/blob/xueyang/speechllm_tts_2402_cfg_audio/nemo/collections/tts/models/speechllm/megatron_t5_speechllm_model.py?ref_type=heads#L1541-1578

**Comment by XuesongYang** [comment]
please hold off reviewing this change. There are still bugs after rebasing. I will mention you to review once I figure them out.

**Review by blisc — APPROVED** [review]
